datafusion/physical-expr/src/aggregate/count_distinct.rs

```diff
@@ -192,20 +285,175 @@ impl Accumulator for DistinctCountAccumulator {
     }
 }
 
+struct DistinctCountGroupsAccumulator<T>
+where
+    T: ArrowPrimitiveType + Send,
+{
+    /// Vector for storing unique values sets for each group index
+    unique_values: Vec<HashSet<Hashable<T::Native>>>,
```

Comment:
I hoped we could store this in something like `HashSet<(usize, T::Native)>`, i.e. we could ammortize the allocations for all the groups instead of creating a hashset for each group.

Reply:
That was my initial intention -- but collecting intermediate states is quite painful when using pure `HashSet<(group, value)>`, and using additional structures for storing chains of values for each group seemed like a memory overhead in this case.

However, I agree that managing single hash table per accumulator makes sense -- maybe the proper way of speeding up accumulator will be switching to native types in regular (current) accumulator first, and then start digging into proper groups accumulator implementation which will provide noticeable boost :thinking:.

Reply:
> maybe the proper way of speeding up accumulator will be switching to native types in regular (current) accumulator first

I'll try this option then, I guess at first -- anyway it was going to be follow up for current version.